src/Middleware/TrustedHostsNetworkResolver.php

```diff
@@ -0,0 +1,510 @@
+            }
+            if (!\is_string($header)) {
+                throw new \InvalidArgumentException('The header is not a string');
+            }
+            if ($type === self::IP_HEADER_TYPE_RFC7239) {
+                continue;
+            }
+
+            throw new \InvalidArgumentException("Not supported IP header type: $type");
+        }
+        if (count($hosts) === 0) {
+            throw new \InvalidArgumentException("Empty hosts not allowed");
+        }
+        $data = [
+            self::DATA_KEY_HOSTS => $hosts,
+            self::DATA_KEY_IP_HEADERS => $ipHeaders,
+            self::DATA_KEY_PROTOCOL_HEADERS => $this->prepareProtocolHeaders($protocolHeaders),
+            self::DATA_KEY_TRUSTED_HEADERS => $trustedHeaders ?? self::DEFAULT_TRUSTED_HEADERS,
+            self::DATA_KEY_HOST_HEADERS => $hostHeaders,
+            self::DATA_KEY_URL_HEADERS => $urlHeaders,
+        ];
+        foreach ([
+                     self::DATA_KEY_HOSTS,
+                     self::DATA_KEY_TRUSTED_HEADERS,
+                     self::DATA_KEY_HOST_HEADERS,
+                     self::DATA_KEY_URL_HEADERS
+                 ] as $key) {
+            $this->checkStringArrayType($data[$key], $key);
+        }
+        foreach ($data[self::DATA_KEY_HOSTS] as $host) {
+            $host = str_replace('*', 'wildcard', $host);        // wildcard is allowed in host
+            if (filter_var($host, FILTER_VALIDATE_DOMAIN) === false) {
+                throw new \InvalidArgumentException("'$host' host is not a domain and not an IP address");
+            }
+        }
+        $new->trustedHosts[] = $data;
+        return $new;
+    }
+
+    private function checkStringArrayType(array $array, string $field): void
+    {
+        foreach ($array as $item) {
+            if (!is_string($item)) {
+                throw new \InvalidArgumentException("$field must be string type");
+            }
+            if (trim($item) === '') {
+                throw new \InvalidArgumentException("$field cannot be empty strings");
+            }
+        }
+    }
+
+    /**
+     * @return static
+     */
+    public function withoutTrustedHosts()
+    {
+        $new = clone $this;
+        $new->trustedHosts = [];
+        return $new;
+    }
+
+    /**
+     * Request's attribute name to which trusted path data is added.
+     *
+     * The list starts with the server and the last item is the client itself.
+     *
+     * @return static
+     * @see getElementsByRfc7239
+     */
+    public function withAttributeIps(?string $attribute)
+    {
+        if ($attribute === '') {
+            throw new \RuntimeException('Attribute should not be empty');
+        }
+        $new = clone $this;
+        $new->attributeIps = $attribute;
+        return $new;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
```

Comment:
Please split this method on some methods.

Reply:
I see that this function of too many lines of code. It can be splitted, but for multiple functions, 4-5 parameters are transported, which is not very nice.